### ISSUE TITLE
SetupAllProperties 2nd property not mocked

### ISSUE THREAD

**Issue by dosfx** [moq/moq4#438]
New to moq, so its probably me but I've got an interface that I'm mocking with 2 get properties of another interface.  When I call SetupAllProperties on the mock, the 2nd get property doesn't appear to get mocked.  All I can think is that it is due to it being the same type?

example:
```csharp
interface IResult
{
    ISubResult Part1 { get; }
    string OtherProperty { get; set; }
    ISubResult Part2 { get; }
}

resultMock = new Mock<IAuthPipeResult> { DefaultValue = DefaultValue.Mock };
resultMock.SetupAllProperties();

resultMock.Object.Part1.Name = "value";  // This is fine
resultMock.Object.Part2.Name = "value";  // Part2 is null, throws exception
```

Currently my workaround is to call ```resultMock.SetupAllProperties();``` and then call ```resultMock.SetupProperty(m => m.Part1.Name);``` but I feel like I must be missing something

Thanks for reading

**Comment by stakx** [comment]
@dosfx, thanks for reporting! You're not missing anything. This is a bug. I'm surprised that noone's run into this before.

`mock.SetupAllProperties` sets up properties recursively. [There's some logic in Moq](https://github.com/moq/moq4/blob/e2662721e13d52b081a2706608cb03e458019279/Source/Mock.cs#L736-L740) to prevent infinite loops during that process (in case of recursive types). The problem here is that this logic works exclusively on types, and not on type instances. When `SetupAllProperties` sets up `Part1`, it notes down that it has mocked something of type `ISubResult`. When it gets to `Part2`, it sees that it is dealing with an `ISubResult` again, and in order to prevent a recursion-based infinite loop, it short-cuts by setting up `Part2` to simply return `null`. This behaviour does not make any sense in your scenario.

The loop detection logic needs to be made more accurate. Until this has happened, you're better off not using `SetupAllProperties` in this particular case. Instead do what you've already figured out yourself and set up the properties individually, e.g.:

```csharp
resultMock.SetupGet(result => result.Part1).Returns(Mock.Of<ISubResult>());
resultMock.SetupGet(result => result.Part2).Returns(Mock.Of<ISubResult>());
```

**Comment by stakx** [comment]
@dosfx, this will be fixed in the next release of Moq (version >4.7.99).

**Comment by dosfx** [comment]
Thank you very much!  Thats great :)
